### PR TITLE
Handle conversion of non-checksummed eth address

### LINKED ISSUE #7407: Automatic conversion of non-checksummed address
**Design:**
convert-checksum

**Spec:**
As a user if I type or paste an Ethereum address in invalid format, Trezor Suite needs to convert the address to a supported format automatically and inform me about it.

When I paste or manually enter valid non-checksummed Ethereum address to the Ethereum Send address field, Trezor Suite converts it to checksummed address

I need to be informed that such automatic conversion happened below the Send address field

**Acceptance criteria**

- Trezor Suite checks the address and converts it to a checksummed version automatically
- Below the recipient’s address on the left side I see blue double check mark which indicates the address was checked as valid and also converted to the checksummed address
- Next to the double check mark I see blue text “Converted to checksum address”
- On the right side I see “Learn more” button which redirects to the KB article ([https://trezor.io/support/a/ethereum-convert-checksum-address](https://trezor.io/support/a/ethereum-convert-checksum-address))
- When I paste or manually enter invalid Ethereum address to the Ethereum Send address field, Trezor Suite shows an error the same way as it does now

### PR DESCRIPTION
**Pull request by enjojoy**
## Description

Based on the conversation in the issue, I've decided to do the following:

1. Check whether the address is a valid ethereum address (already implemented)
2. Check whether the address is used via blockbook
    a. If the address is used but unchecksummed or incorrectly checksummed, then Suite will automatically convert the address to the correct checksummed form and inform the user:
image
    b. if the address haven't been used yet and is not checksummed at all, offer to checksum with a button:
image

...WOP

## Related Issue

Resolve https://github.com/trezor/trezor-suite/issues/7407

### CODE REVIEW

**Comment by enjojoy** [comment]
I'm not sure the double checkmark icon is present in the icons package, I need to add it there as said in the icons [readme](https://github.com/trezor/trezor-suite/blob/develop/suite-common/icons/README.md), so I can use it in the info bottom text, but I'm not sure I have an access to figma. Could you help me to do that please/give access to figma? @AdamSchinzel

**Comment by enjojoy** [comment]
WOP

image
image

**Comment by enjojoy** [comment]
If the address is not checksummed and haven't been used before, but valid (text message to be adjusted)

image

File: packages/components/src/components/form/InputStyles.ts [hunk 1]
```diff
@@ -50,6 +52,8 @@ export const getInputStateBgColor = (
             return theme.backgroundAlertYellowSubtleOnElevation1;
         case 'error':
             return theme.backgroundAlertRedSubtleOnElevation1;
+        case 'info':
+            return theme.backgroundAlertBlueSubtleOnElevation1; //TODO: should we highlight the automatically chacksummed address?
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/InputStyles.ts]: No need to highlight it and also see [this comment](https://github.com/trezor/trezor-suite/pull/12453#discussion_r1609646378) (I would remove this).

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 2]
```diff
@@ -198,9 +206,56 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                //1.If the address is used but unchecksummed or incorrectly checksummed, then Suite will automatically convert the address to the correct checksummed form and inform the user as described in the OP.
+                if (networkType === 'ethereum') {
+                    const params = {
+                        descriptor: value,
+                        coin: 'ethereum',
+                    };
+                    const result = await TrezorConnect.getAccountInfo(params); //TODO handle error
+                    if (result.success && result.payload.history.total !== 0) {
+                        if (!checkAddressCheckSum(value)) {
+                            console.log('used but not checksummed');
```
Comment (AdamSchinzel) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Probably leftover

File: packages/components/src/components/form/BottomText.tsx [hunk 3]
```diff
@@ -39,21 +69,24 @@ interface BottomTextProps {
...
 
     const iconColor: Color | CSSColor = isDisabled
         ? 'iconDisabled'
         : getInputStateTextColor(inputState, theme);
 
     return (
         <Container $inputState={inputState} $isDisabled={isDisabled}>
-            {icon && <Icon name={icon} size="medium" color={iconColor} />}
+            {iconMap[inputState] && (
+                <Icon name={iconMap[inputState]} size="medium" color={iconColor} />
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]: You can define `iconMap[inputState]` as const and then use on both places here.

Plus you should be getting some type errors right now. To fix them I suggest:
```
const iconMap = {
        error: 'warningCircle',
        warning: 'warningCircle',
        info: 'doubleCheck',
} as const satisfies Record<InputState, IconName>;

...

{inputState && iconMap[inputState] && (
       <Icon name={iconMap[inputState]} size="medium" color={iconColor} />
)}
```

File: packages/components/src/components/form/BottomText.tsx [hunk 4]
```diff
@@ -39,21 +69,24 @@ interface BottomTextProps {
     children: ReactNode;
 }
 
-export const BottomText = ({
-    inputState,
-    isDisabled,
-    icon = 'warningCircle',
-    children,
-}: BottomTextProps) => {
+export const BottomText = ({ inputState, isDisabled, children }: BottomTextProps) => {
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]: Remove icon from `BottomTextProps` if you don't use it anymore

File: packages/components/src/components/form/BottomText.tsx [hunk 5]
```diff
@@ -39,21 +69,24 @@ interface BottomTextProps {
...
         : getInputStateTextColor(inputState, theme);
 
     return (
         <Container $inputState={inputState} $isDisabled={isDisabled}>
-            {icon && <Icon name={icon} size="medium" color={iconColor} />}
+            {iconMap[inputState] && (
+                <Icon name={iconMap[inputState]} size="medium" color={iconColor} />
+            )}
+
+            {inputState == 'info' ? <Divider /> : undefined}
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]:
```suggestion
            {inputState === 'info'  && <Divider />}
```

File: packages/components/src/components/form/BottomText.tsx [hunk 6]
```diff
@@ -30,6 +30,36 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
         $isDisabled ? theme.textDisabled : getInputStateTextColor($inputState, theme)};
     ${typography.label}
     animation: ${slideDown} 0.18s ease-in-out forwards;
+    background-color: ${({ $inputState, theme }) =>
+        $inputState == 'info' ? getInputStateBgColor($inputState, theme, 1) : undefined};
+    border-radius: 4px;
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]:
```suggestion
    border-radius: ${borders.radii.xxs};
```

File: packages/components/src/components/form/BottomText.tsx [hunk 7]
```diff
@@ -30,6 +30,36 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
+
+    ${({ $inputState, theme }) =>
+        $inputState === 'info' &&
+        `
+        margin: ${spacingsPx.sm} 0;
+        padding: ${spacingsPx.xs} ${spacingsPx.sm};
+        ::before {
+            content: '';
+            position: absolute;
+            top: -8px;
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]:
```suggestion
            top: -${spacingsPx.xs};
```

File: packages/components/src/components/form/BottomText.tsx [hunk 8]
```diff
@@ -30,6 +30,36 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
         $isDisabled ? theme.textDisabled : getInputStateTextColor($inputState, theme)};
     ${typography.label}
     animation: ${slideDown} 0.18s ease-in-out forwards;
+    background-color: ${({ $inputState, theme }) =>
+        $inputState == 'info' ? getInputStateBgColor($inputState, theme, 1) : undefined};
```
Comment (AdamSchinzel) [inline packages/components/src/components/form/BottomText.tsx]: I think you don't need to modify `getInputStateBgColor` for info case since you are basically hardcoding the output color anyway.

So instead I would prefer to just put `backgroundAlertBlueSubtleOnElevation1` fot both background-color and border-bottom

File: packages/suite/src/components/wallet/InputInfo.tsx [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+import styled from 'styled-components';
+import { useTheme } from 'styled-components';
+import { Button } from '@trezor/components';
+import { LearnMoreButton } from '../suite/LearnMoreButton';
+import { TooltipSymbol } from '../suite';
+import { getInputStateTextColor } from '@trezor/components';
+const Wrapper = styled.div`
+    display: flex;
+    align-items: center;
+    gap: 8px;
```
Comment (AdamSchinzel) [inline packages/suite/src/components/wallet/InputInfo.tsx]:
```suggestion
    gap: ${spacingsPx.xs};
```

Comment (AdamSchinzel) [reply packages/suite/src/components/wallet/InputInfo.tsx]: Also put an empty line between imports and `Wrapper`

File: packages/suite/src/components/wallet/InputInfo.tsx [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+    width: 100%;
+    justify-content: space-between;
+`;
+
+interface InputInfoProps {
+    button?: { onClick: MouseEventHandler<HTMLButtonElement>; text: string } | { url: string };
+    message?: string;
+    tooltip?: { content: string };
```
Comment (AdamSchinzel) [inline packages/suite/src/components/wallet/InputInfo.tsx]: You don't need to put content inside an object. Just do `tooltip?: string;`

File: suite-common/icons/assets/icons/doubleCheck.svg [hunk 11]
```diff
@@ -0,0 +1,3 @@
+<svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 16 16">
+  <path stroke="#0078AC" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="m9.25 5.25-5.5 5.5L1 8m14-2.75-5.5 5.5-1.463-1.463"/>
```
Comment (AdamSchinzel) [inline suite-common/icons/assets/icons/doubleCheck.svg]: Don't put the icon here already with the color, use classic black color not the blue color.

```suggestion
  <path stroke="#171717" stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="m9.25 5.25-5.5 5.5L1 8m14-2.75-5.5 5.5-1.463-1.463"/>
```

File: packages/suite/src/components/wallet/InputInfo.tsx [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+const Wrapper = styled.div`
+    display: flex;
+    align-items: center;
+    gap: 8px;
+    width: 100%;
+    justify-content: space-between;
+`;
+
+interface InputInfoProps {
+    button?: { onClick: MouseEventHandler<HTMLButtonElement>; text: string } | { url: string };
```
Comment (AdamSchinzel) [inline packages/suite/src/components/wallet/InputInfo.tsx]: I know that you put this here for future usage and because you want to approach it in generic way but since you don't use this at all I think you can remove it completely

File: packages/suite/src/components/wallet/InputInfo.tsx [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    message?: string;
```
Comment (AdamSchinzel) [inline packages/suite/src/components/wallet/InputInfo.tsx]:
```suggestion
    message: TranslationKey;
```

File: packages/suite/src/components/wallet/InputInfo.tsx [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+    tooltip?: { content: string };
+}
+export const InputInfo = ({ button, message, tooltip }: InputInfoProps) => {
+    const theme = useTheme();
+    const textColor = getInputStateTextColor('info', theme);
+
+    return (
+        <Wrapper>
+            {message}
```
Comment (AdamSchinzel) [inline packages/suite/src/components/wallet/InputInfo.tsx]:
```suggestion
            <Translation id={message} />
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 15]
```diff
@@ -222,10 +277,28 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                <InputError message={addressError.message} button={getValidationButtonProps()} />
+            );
+        }
+
+        if (addressInfoContent) {
+            const mess = 'Converted to Checksum'; //TODO: add translation
+
+            const tooltip = { content: 'Info about checksum conversion' };
+
+            return <InputInfo message={mess} tooltip={tooltip} />;
```
Comment (AdamSchinzel) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: You can put both values right inside the component no need to put them in separated consts

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 16]
```diff
@@ -198,9 +206,56 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
                 }
             },
             // eth addresses are valid without checksum but Trezor displays them as checksummed
-            checksum: (value: string) => {
-                if (networkType === 'ethereum' && !checkAddressCheckSum(value)) {
-                    return translationString('RECIPIENT_IS_NOT_VALID');
+            used_checksum: async (value: string) => {
```
Comment (AdamSchinzel) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: I don't really get the whole part of this plus you are not returning anything so types are complaining. Can you elaborate what it does and why you even use localStorage etc.

**Review by AdamSchinzel — CHANGES_REQUESTED** [review]

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 16 above)
Comment (AdamSchinzel) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Also using `addressInfoContent` seems to me to not be really needed since you can return some error as a string and if it matches `addressError` then in `getInputState` you can return `info` (so you don't even need to modify input state to void).

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 17]
```diff
@@ -198,9 +206,56 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                        }
+                    }
+                }
+            },
+            checksum: async (value: string) => {
+                if (networkType === 'ethereum') {
+                    const params = {
+                        descriptor: value,
+                        coin: 'ethereum',
+                    };
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]:
```suggestion
            checksum: async (address: string) => {
                if (networkType === 'ethereum') {
                    const params = {
                        descriptor: address,
                        coin: symbol,
                    };
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 18]
```diff
@@ -198,9 +206,56 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                    const result = await TrezorConnect.getAccountInfo(params);
+                    //2. if address is not checksummed at all and not found in blockbook
+                    //offer to checksum width a button
+                    if (result.success && result.payload.history.total == 0) {
+                        if (value == value.toLowerCase()) {
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]:
```suggestion
                        if (address === address.toLowerCase()) {
```
what if it is not checksummed nor lowercase? e.g. its all uppercase 😀  or mixed case but not in the correct checksummed way

**Review by tomasklim — CHANGES_REQUESTED** [review]
Needs some polish but proposed solution looks good

File: packages/components/src/components/form/InputStyles.ts (hunk 1 above)
Comment (enjojoy) [reply packages/components/src/components/form/InputStyles.ts]: removed

File: packages/components/src/components/form/BottomText.tsx (hunk 8 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx]: Yep, I agree and fixed

File: packages/components/src/components/form/BottomText.tsx (hunk 3 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx]: thank you, fixed
https://github.com/trezor/trezor-suite/pull/12453/commits/090f4526bb963c42a2aafa667b013c68f1d1d5d7

File: packages/suite/src/components/wallet/InputInfo.tsx (hunk 12 above)
Comment (enjojoy) [reply packages/suite/src/components/wallet/InputInfo.tsx]: oki, yep I thought it might be useful in the future, but we can add it later in case we need it. removed

File: packages/suite/src/components/wallet/InputInfo.tsx (hunk 10 above)
Comment (enjojoy) [reply packages/suite/src/components/wallet/InputInfo.tsx]: true, i changed it to tooltipContent and now passing only a string

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 16 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: > I don't really get the whole part of this plus you are not returning anything so types are complaining. Can you elaborate what it does and why you even use localStorage etc.

I will remake the solution, it was my first iteration that worked, but i agree that it doesn't look good. I won't use localStorage either.

File: suite-common/icons/assets/icons/doubleCheck.svg (hunk 11 above)
Comment (enjojoy) [reply suite-common/icons/assets/icons/doubleCheck.svg]: fixed, also  on the info icon

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 18 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: If it's all uppercase, it can be treated the same way as all lowercase. But in case of mixed case (but wrong checkdummed), it should be just invalid, because it means that the address was affected by a typo. By automatically checksumming such an address we put the user under a risk of using a non-existent address. Hence the user should double-check and fix it himself.

Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: If it's incorrectly checksummed in mixed case it's catched much earlier than the checksum validation. So no matter if the account is used, the uncorrectly checksummed address will be not valid. Here: 

            valid: (value: string) => {
                if (!isAddressValid(value, symbol)) {
                    return translationString('RECIPIENT_IS_NOT_VALID');
                }
            },

Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: If it's all upprercase it's a bit tricky. If the x in 0x is uppercase too, it's invalid, if the x is lowercase but the rest of the address is all uppercase, it gets converted to checksum. Is that the desired behavior?

File: packages/components/src/components/form/BottomText.tsx [hunk 19]
```diff
@@ -30,30 +30,63 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
-    icon = 'warningCircle',
-    children,
-}: BottomTextProps) => {
+export const BottomText = ({ inputState, isDisabled, children }: BottomTextProps) => {
     const theme = useTheme();
+    const iconMap = {
+        error: 'warningCircle',
+        warning: 'warningCircle',
+        info: 'doubleCheck',
+    } as const satisfies Record<InputState, IconName>;
```
Comment (jvaclavik) [inline packages/components/src/components/form/BottomText.tsx]: We can use default icon for each input state, but we should be able to choose any icon with different prop.

Please take a look on component Warning (`packages/components/src/components/Warning/Warning.tsx`) and refactor it in the same style.

Screenshot 2024-06-07 at 14 24 11

Screenshot 2024-06-07 at 14 25 58

File: packages/components/src/components/form/BottomText.tsx [hunk 20]
```diff
@@ -30,30 +30,63 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
+            top: -${spacingsPx.xs};
+            left: 20px;
+            transform: translateX(-50%);
+            width: 0;
+            height: 0;
+            border-left: 8px solid transparent; 
+            border-right: 8px solid transparent; 
+            border-bottom: 8px solid ${theme.backgroundAlertBlueSubtleOnElevation1}; 
+        }
+        `}
```
Comment (jvaclavik) [inline packages/components/src/components/form/BottomText.tsx]: I believe this shouldn't be here, because `BottomText` is generic component in the design system. 

I suggest – if you want to use some custom component – just pass `bottomText`: (as custom ReactNode not string) to `Input` and style it on your own outside of the design system.

Something like:

```tsx
<Input 
  ...
  bottomText={<CustomComponent/>}
/>
```

File: packages/components/src/components/form/BottomText.tsx [hunk 21]
```diff
@@ -30,30 +30,63 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
+    align-self: stretch;
+    width: 2px;
+    background-color: ${({ theme }) => theme.backgroundSurfaceElevation1};
+    margin: -${spacingsPx.xs} ${spacingsPx.xs};
 `;
 
 interface BottomTextProps {
     inputState?: InputState;
     isDisabled?: boolean;
-    icon?: IconName;
```
Comment (jvaclavik) [inline packages/components/src/components/form/BottomText.tsx:38]: Why do you need it? We want to keep this prop here.

File: packages/components/src/components/form/BottomText.tsx [hunk 22]
```diff
@@ -30,30 +30,63 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
         : getInputStateTextColor(inputState, theme);
 
     return (
         <Container $inputState={inputState} $isDisabled={isDisabled}>
-            {icon && <Icon name={icon} size="medium" color={iconColor} />}
+            {inputState && iconMap[inputState] && (
+                <Icon name={iconMap[inputState]} size="medium" color={iconColor} />
+            )}
+
+            {inputState === 'info' && <Divider />}
```
Comment (jvaclavik) [inline packages/components/src/components/form/BottomText.tsx]: This shouldn't be dependent on the same. Also if you need it create BottomText as a custom component or add generic prop `hasDivider`.

**Review by jvaclavik — CHANGES_REQUESTED** [review]

File: packages/components/src/components/form/BottomText.tsx (hunk 22 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx]: I see what you mean. I will rather create a new custom component and pass it into BottomText & leave BottomText untouched besides the Warning-like refactoring, so we can get rid of the icon when needed & have default re writable icons.

File: packages/components/src/components/form/BottomText.tsx (hunk 21 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx:38]: Returned

File: packages/components/src/components/form/BottomText.tsx (hunk 22 above)
Comment (jvaclavik) [reply packages/components/src/components/form/BottomText.tsx]: Sounds good 👍

File: packages/components/src/components/form/BottomText.tsx (hunk 19 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx]: This behavior is introduced now in BottomText component

```
 let iconName;

    if (icon) {
        iconName = icon;
    } else if (inputState && iconMap[inputState]) {
        iconName = iconMap[inputState];
    }

    return (
        <Container $inputState={inputState} $isDisabled={isDisabled}>
            {withIcon && iconName && <Icon name={iconName} size="medium" color={iconColor} />}
            
```

File: packages/components/src/components/form/BottomText.tsx (hunk 20 above)
Comment (enjojoy) [reply packages/components/src/components/form/BottomText.tsx]: This is removed and a custom component is introduced

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 23]
```diff
@@ -222,10 +286,29 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+        }
+
+        return addressBottomText;
+    };
+
     return (
         <Container>
             <Input
-                inputState={inputState}
+                inputState={inputState()}
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: nit:
```suggestion
                getInputState={inputState()}
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 24]
```diff
@@ -222,10 +286,29 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+            return (
+                <InputError message={addressError.message} button={getValidationButtonProps()} />
+            );
+        }
+
+        if (addressChecksumInfo) {
+            return (
+                <InputChecksumInfo
+                    message="TR_CHECKSUM_CONVERSION_INFO"
+                    tooltipContent="Info about checksummed address"
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Untranslatable
```suggestion
                    tooltipContent="SOME_TRANSLATION_KEY"
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 25]
```diff
@@ -198,9 +226,45 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                            setValue(inputName, toChecksumAddress(address), {
+                                shouldValidate: true,
+                            });
+                        }
+                    }
+                    //2. if address is not checksummed at all and not found in blockbook
+                    //offer to checksum width a button
+                    if (result.success && result.payload.history.total == 0) {
+                        if (address == address.toLowerCase()) {
+                            return 'Not used and not checksummed at all';
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Untranslatable

File: packages/components/src/components/form/BottomText.tsx [hunk 26]
```diff
@@ -32,29 +40,35 @@ export const Container = styled.div<{ $inputState?: InputState; $isDisabled?: bo
...
         : getInputStateTextColor(inputState, theme);
 
+    let iconName;
+
+    if (icon) {
+        iconName = icon;
+    } else if (inputState && iconMap[inputState]) {
+        iconName = iconMap[inputState];
+    }
+
```
Comment (tomasklim) [inline packages/components/src/components/form/BottomText.tsx]:
```suggestion
    const iconName = icon || (inputState && iconMap[inputState]);
```

File: packages/components/src/components/form/BottomText.tsx [hunk 27]
```diff
@@ -9,6 +9,14 @@ import { InputState } from './inputTypes';
 
 export const BOTTOM_TEXT_MIN_HEIGHT = 26; // 1 line of text + top padding
 
+interface BottomTextProps {
```
Comment (tomasklim) [inline packages/components/src/components/form/BottomText.tsx]: nit: We put interfaces above components

File: packages/components/src/components/form/Input/Input.tsx [hunk 28]
```diff
@@ -198,7 +198,11 @@ const Input = ({
             </InputWrapper>
 
             {bottomText && (
-                <BottomText inputState={inputState} isDisabled={isDisabled}>
+                <BottomText
+                    withIcon={inputState === 'error' || inputState === 'warning'}
```
Comment (tomasklim) [inline packages/components/src/components/form/Input/Input.tsx]: nit:
```suggestion
                    withIcon={['error', 'warning'].includes(inputState)}
```

File: packages/suite/src/components/wallet/InputChecksumInfo.tsx [hunk 29]
```diff
@@ -0,0 +1,74 @@
...
+    gap: ${spacingsPx.xs};
+    width: 100%;
+    justify-content: space-between;
+    background-color: ${({ theme }) => theme.backgroundAlertBlueSubtleOnElevation1};
+    border-radius: ${borders.radii.xxs};
+    position: relative;
+    margin: ${spacingsPx.xxs} 0 ${spacingsPx.sm} 0;
+    padding: ${spacingsPx.xs} ${spacingsPx.sm};
+    ${({ theme }) =>
+        `
```
Comment (tomasklim) [inline packages/suite/src/components/wallet/InputChecksumInfo.tsx]:
```suggestion
    ${({ theme }) => css`
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 30]
```diff
@@ -47,10 +48,32 @@ interface AddressProps {
...
+
+    return { lastTwoValues, addValue };
+};
+
 export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
     const [addressDeprecatedUrl, setAddressDeprecatedUrl] = useState<
         AddressDeprecatedUrl | undefined
     >(undefined);
+    type AddressChecksumInfo = { text: string } | undefined;
+    const [addressChecksumInfo, setAddressChecksumInfo] = useState<AddressChecksumInfo>(undefined);
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Why wrapping `text` into object?
`type` should not be inside component
```suggestion
const [addressChecksumInfo, setAddressChecksumInfo] = useState<string | undefined>();
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 31]
```diff
@@ -198,9 +226,45 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                    const result = await TrezorConnect.getAccountInfo(params); //TODO handle error
+                    if (result.success && result.payload.history.total !== 0) {
+                        if (!checkAddressCheckSum(address)) {
+                            setValue(inputName, toChecksumAddress(address), {
+                                shouldValidate: true,
+                            });
+                        }
+                    }
+                    //2. if address is not checksummed at all and not found in blockbook
+                    //offer to checksum width a button
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]:
```suggestion
                    // 2. If the address is not checksummed at all and not found in Blockbook,
                    // offer to checksum it with a button.
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 32]
```diff
@@ -198,9 +226,45 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                    if (result.success && result.payload.history.total == 0) {
+                        if (address == address.toLowerCase()) {
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: use `===` + you can wrap it into ` if (result.success) {`
```suggestion
                    if (result.success) {
                        if (result.payload.history.total !== 0 && !checkAddressCheckSum(address)) {
                            setValue(inputName, toChecksumAddress(address), {
                                shouldValidate: true,
                            });
                        }
                  
                        if (result.payload.history.total === 0 && address === address.toLowerCase()) {
```

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 33]
```diff
@@ -198,9 +226,45 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                        updatedValues[1] &&
+                        updatedValues[0] !== updatedValues[1] &&
+                        toChecksumAddress(updatedValues[0]) === updatedValues[1]
+                    ) {
+                        setAddressChecksumInfo({
+                            text: 'Address converted to checksum format.',
+                        });
+                    } else {
+                        setAddressChecksumInfo(undefined);
+                    }
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Why is this required? What if we just put 
this
```
                 setAddressChecksumInfo({
                            text: 'Address converted to checksum format.',
                        });
```
here

```
if (result.success && result.payload.history.total !== 0) {
                        if (!checkAddressCheckSum(address)) {
                            setValue(inputName, toChecksumAddress(address), {
                                shouldValidate: true,
                            });
                            setAddressChecksumInfo({
                                text: 'Address converted to checksum format.',
                            });
                        }
                    }
```
and reset `addressChecksumInfo` `onChange` of address if it is set

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 34]
```diff
@@ -198,9 +226,45 @@ export const Address = ({ output, outputId, outputsCount }: AddressProps) => {
...
+                    //they are not the same but the second one is checksum of the first one, in this case show the info that the address was automatically checksummed
+                    if (
+                        updatedValues &&
+                        updatedValues[0] &&
+                        updatedValues[1] &&
+                        updatedValues[0] !== updatedValues[1] &&
+                        toChecksumAddress(updatedValues[0]) === updatedValues[1]
+                    ) {
+                        setAddressChecksumInfo({
+                            text: 'Address converted to checksum format.',
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: This string is not used at all, why not calling this `addressChecksumInfo` -> `hasAddressBeenChecksummed` boolean?

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx [hunk 35]
```diff
@@ -47,10 +48,32 @@ interface AddressProps {
...
+            }
+
+            return newValues;
+        });
+
+        return newValues;
+    };
+
+    return { lastTwoValues, addValue };
+};
```
Comment (tomasklim) [inline packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: I think this hook is not needed, info in some comment below. Moreover, it should not be part of this file

**Review by tomasklim — CHANGES_REQUESTED** [review]
I think commits are little bit chaotic. Could you please reset them and commit it again in less / better described commits?

I dont understand the `useLastTwoValues` logic but maybe I am missing something

File: packages/components/src/components/form/Input/Input.tsx (hunk 28 above)
Comment (enjojoy) [reply packages/components/src/components/form/Input/Input.tsx]: nice!!

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 23 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: I'm not sure I understand why this rename is needed

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 30 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: I realized this text is not used, it will be a boolean

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 34 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Yep, just did this myself too :D

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 33 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: You are actually right. 

The problem was when the address was changed through toChecksumAddress(), it also triggered the validation logic again and the information bubble disappeared. Thats why I introduced that non very elegant logic. To be completely honest I didn't know that onChange is triggered only when the input is changed through the keyboard or paste, but not when it changed programatically with the toChecksum...() func.  This solves everything and that logic is not needed.

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 35 above)
Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: yes, removed

File: packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx (hunk 23 above)
Comment (tomasklim) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: you are right, i wanted to comment it like this

```suggestion
                inputState={getInputState()}
```

Comment (tomasklim) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: ah there is conflict in names.

However, `inputState` is a function so it should have name which stresses that it is a function and not a variable

Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: makes sense, I will think of another name

Comment (enjojoy) [reply packages/suite/src/views/wallet/send/components/Outputs/components/Address/index.tsx]: Renamed to getInputErrorState()
